### PR TITLE
Add containSubstring expression function to check for substring in a string

### LINKED ISSUE #2804: Add a function in the DataPrepper Expressions to check if a string contains another string
**Is your feature request related to a problem? Please describe.**
Add a function in the DataPrepper Expressions to check if a string contains another string. This is useful to check if a field in the event has a particular prefix/suffix/substring in it.

**Describe the solution you'd like**
Proposal is to add `containsSubstring` function to DataPrepper expressions which can take string literal or a json pointer to check if a string contains another string as it's substring. 
The function signature is as follows
```
boolean containsSubstring(String string1, String string2)
```
It returns true if the string in the `string2` is a substring of `string1`, Otherwise it returns false.

It can be used in the following 4 ways
```
containsSubstring(/jsonPointer, "stringLiteral")
containsSubstring("stringLiteral", /jsonPointer)
containsSubstring("stringLiteral1", "stringLiteral2")
containsSubstring(/jsonPointer1, /jsonPointer2)
```

**Describe alternatives you've considered (Optional)**

**Additional context**

### PR DESCRIPTION
**Pull request by kkondaka**
### Description
Add containSubstring expression function to check for substring in a string.
The function signature is as follows

```
boolean containsSubstring(String string1, String string2)
```

It returns true if the string in the string2 is a substring of string1, Otherwise it returns false.

It can be used in the following 4 ways
```
containsSubstring(/jsonPointer, "stringLiteral")
containsSubstring("stringLiteral", /jsonPointer)
containsSubstring("stringLiteral1", "stringLiteral2")
containsSubstring(/jsonPointer1, /jsonPointer2)
```

Resolves #2804 
 
### Issues Resolved
Resolves #2804 
 
### Check List
- [ X] New functionality includes testing.
- [x] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ X] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/ContainsSubstringExpressionFunction.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+import java.util.function.Function;
+
+@Named
+public class ContainsSubstringExpressionFunction implements ExpressionFunction {
+    public String getFunctionName() {
+        return "containsSubstring";
+    }
+
+    public Object evaluate(final List<Object> args, Event event, Function<Object, Object> convertLiteralType) {
+        if (args.size() != 2) {
```
Comment (cmanning09) [inline data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/ContainsSubstringExpressionFunction.java:20]: Take it or leave it: `2` appears to be a repeated value here. I would recommend creating a constant in this class: `NUMBER_OF_ARGS`

File: data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/ContainsSubstringExpressionFunction.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+            String str = (String) arg;
+            if (str.charAt(0) == '"') {
+                strArgs[i] = str.substring(1, str.length()-1);
+            } else if (str.charAt(0) == '/') {
+                Object obj = event.get(str, Object.class);
+                if (obj == null) {
+                    return false;
+                }
+                if (!(obj instanceof String)) {
+                    throw new RuntimeException("containsSubstring() takes only string type arguments");
```
Comment (cmanning09) [inline data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/ContainsSubstringExpressionFunction.java:38]: For all of the exceptions thrown in this class, let's be more specific here and provide our users with the parameter that result in the exception. This will help them with debugging any failures in their pipeline configuration or data.

```
throw new RuntimeException(String.format("containsSubstring() takes only string type arguments. "%s" is not of type string", arg);
```

File: data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/ContainsSubstringExpressionFunctionTest.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    void testContainsSubstringMultipleSubstrings() {
+        containsSubstringExpressionFunction = createObjectUnderTest();
+        String testString = RandomStringUtils.randomAlphabetic(10);
+        for (int i = 1; i < testString.length() - 1; i++) {
+            assertThat(containsSubstringExpressionFunction.evaluate(List.of("\""+testString+"\"", "\""+testString.substring(0, i)+"\""), testEvent, testFunction), equalTo(true));
+        }
+    }
+
+    @Test
+    void testInvalidContainsSubstring() {
```
Comment (cmanning09) [inline data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/ContainsSubstringExpressionFunctionTest.java:77]: Take it or leave it: These tests may be simplified by using `@ParameterizedTest`. That would require refactoring to ensure the event is created statically as well. This would reduce the number assertions and allow anyone to extend the test by focusing just on invalid values instead of writing the repeated line.

File: data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/ContainsSubstringExpressionFunctionTest.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        assertThat(containsSubstringExpressionFunction.evaluate(List.of("\"abcde\"", "\"xyz\""), testEvent, testFunction), equalTo(false));
+        assertThat(containsSubstringExpressionFunction.evaluate(List.of("/unknown", "xyz"), testEvent, testFunction), equalTo(false));
+        assertThat(containsSubstringExpressionFunction.evaluate(List.of("\"abcde\"", "/unknown"), testEvent, testFunction), equalTo(false));
+    }
+
+    @Test
+    void testContainsSubstringMultipleSubstrings() {
+        containsSubstringExpressionFunction = createObjectUnderTest();
+        String testString = RandomStringUtils.randomAlphabetic(10);
+        for (int i = 1; i < testString.length() - 1; i++) {
```
Comment (dlvenable) [inline data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/ContainsSubstringExpressionFunctionTest.java:71]: I think this could be refactored to be an `@ParameterizedTest`. I generally recommend that approach when possible as it is clearer that it is testing different cases rather than testing collections (which naturally need a loop).

File: data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/ContainsSubstringExpressionFunction.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        String[] strArgs = new String[2];
+        for (int i = 0; i < 2; i++) {
+            Object arg = args.get(i);
+            if (!(arg instanceof String)) {
+                throw new RuntimeException("containsSubstring() takes only string type arguments");
+            }
+            String str = (String) arg;
+            if (str.charAt(0) == '"') {
+                strArgs[i] = str.substring(1, str.length()-1);
+            } else if (str.charAt(0) == '/') {
```
Comment (dlvenable) [inline data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/ContainsSubstringExpressionFunction.java:32]: We really should think about how to get the expression framework to make the decision here, or at least have functions that can decide this. We've seen this in other functions as well.

**Review by dlvenable — DISMISSED** [review]
Please also create a documentation issue in the documentation page in GitHub.

File: data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/ContainsSubstringExpressionFunction.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+package org.opensearch.dataprepper.expression;
+
+import org.opensearch.dataprepper.model.event.Event;
+import java.util.List;
+import javax.inject.Named;
+import java.util.function.Function;
+
+@Named
+public class ContainsSubstringExpressionFunction implements ExpressionFunction {
```
Comment (cmanning09) [inline data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/ContainsSubstringExpressionFunction.java:14]: I put this question on the issue this PR is for. but wanted to repeat this here: 

I think we should make a decision on how'd we want our expressions to look long term now. Do we want common methods or specific type methods. For example: Do we want to support a single function called `contains` which can support only `strings` and `jsonPointers` referencing `strings` in this first iteration.  But then can be extended to support other data types like `sets`, `maps`, etc. in future revisions. Or do we want to have separate functions for  different data types? I see the benefits and draw backs of both. Have we decided on how we will approach this?

File: data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/ContainsSubstringExpressionFunctionTest.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    @Test
+    void testContainsSubstringBasic() {
+        containsSubstringExpressionFunction = createObjectUnderTest();
+        assertThat(containsSubstringExpressionFunction.evaluate(List.of("\"abcde\"", "\"abcd\""), testEvent, testFunction), equalTo(true));
+        assertThat(containsSubstringExpressionFunction.evaluate(List.of("/"+testKey, "/"+testKey2), testEvent, testFunction), equalTo(true));
+        assertThat(containsSubstringExpressionFunction.evaluate(List.of("\""+testValue+"\"", "/"+testKey2), testEvent, testFunction), equalTo(true));
+        assertThat(containsSubstringExpressionFunction.evaluate(List.of("/"+testKey, "\""+testValue2+"\""), testEvent, testFunction), equalTo(true));
+        assertThat(containsSubstringExpressionFunction.evaluate(List.of("\"abcde\"", "\"xyz\""), testEvent, testFunction), equalTo(false));
+        assertThat(containsSubstringExpressionFunction.evaluate(List.of("/unknown", "xyz"), testEvent, testFunction), equalTo(false));
+        assertThat(containsSubstringExpressionFunction.evaluate(List.of("\"abcde\"", "/unknown"), testEvent, testFunction), equalTo(false));
```
Comment (cmanning09) [inline data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/ContainsSubstringExpressionFunctionTest.java:64]: Can we add a test for a jsonPointer referencing a null value.

File: docs/expression_syntax.md [hunk 8]
```diff
@@ -184,6 +184,10 @@ Currently, the following functions are supported
    - takes one String literal as argument. This is the key to lookup in the event's metadata. If the key contains "/", then recursive lookup into the metadata attributes is done.
    - returns the value corresponding to the argument (key) passed. Value can be of any type.
    For example, if metadata contains {"key1": "value2", "key2": 10}, then `getMetadata("key1")` returns "value2", and `getMetadata("key2")` return 10.
+ * `containsSubstring()`
```
Comment (dlvenable) [inline docs/expression_syntax.md:187]: This is still `containsSubstring`.

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/ContainsSubstringExpressionFunction.java (hunk 6 above)
Comment (kkondaka) [reply data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/ContainsSubstringExpressionFunction.java]: We discussed this in a meeting and agreed to rename the function to `contains` to potentially extend the function to other types in future.

File: data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/ContainsSubstringExpressionFunctionTest.java (hunk 3 above)
Comment (kkondaka) [reply data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/ContainsSubstringExpressionFunctionTest.java]: I don't think that would work because the type is a list. Which means I have to write another `methodSource` that generates a list because `valueSource` does not take a list. If I end up writing a method, then I don't think there is much difference between the two tests.

File: data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/ContainsSubstringExpressionFunctionTest.java (hunk 7 above)
Comment (kkondaka) [reply data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/ContainsSubstringExpressionFunctionTest.java]: I am not sure there is any value in such test. `null` is not an instance of String. So, there is no difference in the test with an integer and null, as they both testing the same code path.

**Review by dlvenable — APPROVED** [review]

**Review by cmanning09 — APPROVED** [review]
